### PR TITLE
Add console to bridge

### PR DESCRIPTION
**Pull request by felixmulder**
This PR does two things:

1. Adds the ability to use `console` from within an sbt-session that is using the dotty-bridge
1. Creates an aggregated subproject in the build file so that we may publish the correct snapshot artifacts using `sbt dotty-core/publishLocal` or to deploy real snapshots: `sbt dotty-core/publish`

### CODE REVIEW

File: bridge/src/main/scala/xsbt/ConsoleInterface.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    classpathString: String,
+    log: Logger
+  ): Array[String] = args
+
+  def run(args: Array[String],
+          bootClasspathString: String,
+          classpathString: String,
+          initialCommands: String,
+          cleanupCommands: String,
+          loader: ClassLoader,
```
Comment (smarter) [inline bridge/src/main/scala/xsbt/ConsoleInterface.scala]: `loader` isn't passed to the REPL, is that a TODO? If yes that should be noted in a comment

File: bridge/src/main/scala/xsbt/ConsoleInterface.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import dotty.tools.dotc.repl.REPL
+import dotty.tools.dotc.repl.REPL.Config
+
+class ConsoleInterface {
+  def commandArguments(
+    args: Array[String],
+    bootClasspathString: String,
+    classpathString: String,
+    log: Logger
+  ): Array[String] = args
```
Comment (smarter) [inline bridge/src/main/scala/xsbt/ConsoleInterface.scala]: Did you figure out what this method is for? Could you add a comment explaining why it makes sense to return just `args` here?

File: bridge/src/main/scala/xsbt/ConsoleInterface.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    cleanupCmds: List[String],
+    boundVals: Array[(String, Any)]
+  ) = new Config {
+    override val initialCommands: List[String] = initCmds
+    override val cleanupCommands: List[String] = cleanupCmds
+    override val boundValues: Array[(String, Any)] = boundVals
+  }
+
+  def customRepl(cfg: Config): REPL = new REPL {
+    override lazy val config = cfg
```
Comment (smarter) [inline bridge/src/main/scala/xsbt/ConsoleInterface.scala]: Why lazy?

Comment (felixmulder) [inline bridge/src/main/scala/xsbt/ConsoleInterface.scala]: Must be lazy to override lazy val ;)

File: project/Build.scala [hunk 4]
```diff
@@ -349,6 +351,12 @@ object DottyInjectedPlugin extends AutoPlugin {
      )
    )
 
+   lazy val `dotty-core` = project
```
Comment (smarter) [inline project/Build.scala]: I think the root project should be the one that aggregates the others (like in Ensime: https://github.com/ensime/ensime-server/blob/2.0/project/EnsimeBuild.scala#L243) but I don't know if that requires doing something special to get `compile`, `run`, etc to do the correct thing

File: bridge/src/main/scala/xsbt/ConsoleInterface.scala (hunk 2 above)
Comment (felixmulder) [inline bridge/src/main/scala/xsbt/ConsoleInterface.scala]: In the scalac version it looks like it's just setting up the state of the compiler during this method call. Not sure if that makes sense for dotty...

File: bridge/src/main/scala/xsbt/ConsoleInterface.scala (hunk 1 above)
Comment (felixmulder) [inline bridge/src/main/scala/xsbt/ConsoleInterface.scala]: Good point, it isn't necessary to pass the loader - but if a modified loader is passed somehow, it makes sense that it be propagated to the REPL. Will add that in the morning :) thanks!

File: project/Build.scala (hunk 4 above)
Comment (felixmulder) [inline project/Build.scala]: That is the intention of `dotty-core`, once @sjrd is done with the restructuring this could simply be renamed to `dotty`.

File: src/dotty/tools/dotc/repl/CombinedClassLoader.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
+package dotty.tools
+package dotc
+package repl
+
+import scala.util.Try
+
+class CombinedClassLoader(loaders: List[ClassLoader]) extends ClassLoader {
```
Comment (felixmulder) [inline src/dotty/tools/dotc/repl/CombinedClassLoader.scala]: Yes, this guy needs to go away - not sure how yet though. Every other solution seems to create a circular dependency on sbt 👎

Comment (smarter) [inline src/dotty/tools/dotc/repl/CombinedClassLoader.scala]: Why is this needed? Does scalac do something similar?

Comment (felixmulder) [inline src/dotty/tools/dotc/repl/CombinedClassLoader.scala]: They just set it as the class loader - but that doesn't work for us. So I decided that it would be a good idea to merge the ClassLoaders and look through both of them when searching for classes.

But I don't like this solution - if you have a better one - I'm all ears.

Comment (smarter) [inline src/dotty/tools/dotc/repl/CombinedClassLoader.scala]: Why doesn't it work for us? Can't we replace `parentClassLoader` in `val classLoader` with the classloader given by sbt?

Comment (smarter) [inline src/dotty/tools/dotc/repl/CombinedClassLoader.scala]: In general, I would advise against making a custom subclass of ClassLoader because it can very easily not do what you think it should do, and it's hard to debug, but I don't know enough about the situation here to say for sure.

**Comment by felixmulder** [comment]
This feels stable enough for review now - @DarkDimius

File: src/dotty/tools/dotc/repl/CompilingInterpreter.scala [hunk 6]
```diff
@@ -153,8 +155,10 @@ class CompilingInterpreter(out: PrintWriter, ictx: Context) extends Compiler wit
   */
   /** class loader used to load compiled code */
   val classLoader: ClassLoader = {
-    val parent = new URLClassLoader(compilerClasspath.toArray, parentClassLoader)
-    new AbstractFileClassLoader(virtualDirectory, parent)
+    lazy val parent = new URLClassLoader(compilerClasspath.toArray,
+                                         classOf[Interpreter].getClassLoader)
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/repl/CompilingInterpreter.scala]: why `Interpreter` instead of `CompilingInterpreter`?

Comment (felixmulder) [inline src/dotty/tools/dotc/repl/CompilingInterpreter.scala]: Good question - I think this was originally written by @odersky

File: project/Build.scala [hunk 7]
```diff
@@ -207,11 +209,15 @@ object DottyBuild extends Build {
         "org.scala-sbt" % "api" % sbtVersion.value % "test",
         "org.specs2" %% "specs2" % "2.3.11" % "test"
       ),
-      version := "0.1.1-SNAPSHOT",
-      // The sources should be published with crossPaths := false, the binaries
-      // are unused so it doesn't matter.
+      version :=
+        "0.1.1-" + VersionUtil.commitDate + "-" + VersionUtil.gitHash + "-SNAPSHOT",
```
Comment (DarkDimius) [inline project/Build.scala]: I'm not sure if it is a good idea to publish hundreds of distinct snapshots on maven and ask users to specify git-hash by hand. I'd prefer to publish one snapshot and have version information reported by REPL. Something similar to https://github.com/hmrc/sbt-git-stamp may be in handy for implementing this.

**Comment by DarkDimius** [comment]
LGTM, otherwise.
I prefer to fast-track this PR, as publishing snapshots depends on it.
